### PR TITLE
Rewrite documentation URL mapping to use the TOC

### PR DESCRIPTION
**Pull request by ekohl**
Upstream builds various guides and each guide is built as a single large page. That page has many chapters. Technically every guide is also built into 3 flavors (foreman-deb, foreman-el, katello).

Downstream builds various guides, which mostly come from upstream. Most guides are a simple translation where they're downcased, but some are rewritten to something else.

A bigger difference is that each guide is built into multiple pages. This means we need to find the right page containing the chapter. The Table of Content contains a complete mapping with this information, so it can be looked up dynamically.

If this still can't be found, we use the html-single where everything is built into a single page. We then just hope it exists. Otherwise the browser will just show the whole page. This is an unexpected case, so a warning is logged.

The result is that a lot less maintenance is needed, because in most cases it will "just work". A downside is that it becomes hard to verify links are correct. It becomes needed to crawl through all of the source code for any documentation links, similar to how we crawl through it for translations.

Includes https://github.com/RedHatSatellite/foreman_theme_satellite/pull/59 and https://github.com/RedHatSatellite/foreman_theme_satellite/pull/60.

### CODE REVIEW

File: lib/foreman_theme_satellite/documentation-toc.json
Comment (ekohl) [inline lib/foreman_theme_satellite/documentation-toc.json:1]: The CI jobs needs to be updated to reflect this

File: lib/foreman_theme_satellite/documentation.rb [hunk 1]
```diff
@@ -77,28 +77,44 @@ module Documentation
...
+    # Guide mapping. If no entry is present, it should be assumed that it can
+    # be downcased
+    DOCS_GUIDE_MAPPING = {
+      'Managing_Configurations_Ansible' => 'managing_configurations_using_ansible_integration',
+    }.freeze
+
+    # An upstream chapter mapping, in case downstream another chapter should be
+    # used. The top level key is the upstream guide name where the value is a
+    # mapping of upstream chapter to downstream mapping
+    DOCS_GUIDE_CHAPTER_MAPPING = {
```
Comment (ekohl) [inline lib/foreman_theme_satellite/documentation.rb:103]: I think it's questionable if we really want this, but I kept it in since we had this.

File: lib/tasks/foreman_theme_satellite_tasks.rake [hunk 2]
```diff
@@ -44,13 +44,19 @@ namespace :foreman_theme_satellite do
...
     all_links = ForemanThemeSatellite::Documentation::USER_GUIDE_DICTIONARY
                 .merge(ForemanThemeSatellite::Documentation::PLUGINS_DOCUMENTATION)
-                .merge(ForemanThemeSatellite::Documentation.flat_docs_guides_links)
 
     failed = all_links.filter { |_key, doc_address| doc_address.include?('/html/') && !checker.test_link(doc_address) }
 
+    DOCS_GUIDE_CHAPTER_MAPPING.each do |guide, chapters|
+      chapters.keys.each do |chapter|
+        url = ForemanThemeSatellite::Documentation.docs_url(guide, chapter)
+        failed["/#{guide}/#{chapter}"] = url if url.include?('/html-single/')
```
Comment (ekohl) [inline lib/tasks/foreman_theme_satellite_tasks.rake]: Not sure if this reports correctly.

**Review by ekohl — COMMENTED** [review]
@ShimShtein this builds on your TOC work. The aim is to reduce maintenance burden: it makes PRs like https://github.com/RedHatSatellite/foreman_theme_satellite/pull/45 not needed, meaning there's a greater change it'll "just work".

File: lib/foreman_theme_satellite/documentation.rb [hunk 3]
```diff
@@ -77,28 +77,38 @@ module Documentation
...
-      key_values = []
-      source.map do |k, v|
-        key = "#{prefix}/#{k}"
-        if v.is_a?(Hash)
-          key_values.append(*nested_to_flat_k_v(key, v))
+        if (page = DOCS_GUIDE_PAGES[downstream_guide]&.find { |_page, chapters| chapters.include?(downstream_chapter) }&.first)
+          "#{root}/#{downstream_guide}/#{page}##{downstream_chapter}"
         else
-          key_values.append([key, v])
+          Rails.logger.warn("Could not find page for chapter '#{downstream_chapter}' in guide '#{downstream_guide}'")
```
Comment (ekohl) [inline lib/foreman_theme_satellite/documentation.rb]: The rake task fails because it hasn't initialized Rails so `Rails.logger` is `nil`. I wonder what to do. It may work to use
```suggestion
          Rails.logger&.warn("Could not find page for chapter '#{downstream_chapter}' in guide '#{downstream_guide}'")
```

Another approach is to move `Rails.logger` to the caller.

And we can make the validation initialize Rails.

Thoughts?

File: lib/tasks/foreman_theme_satellite_tasks.rake [hunk 4]
```diff
@@ -14,26 +14,11 @@ namespace :test do
...
-
-    Rake::Task['rubocop_foreman_theme_satellite'].invoke
-  end
-
   desc 'Validate documentation links against given TOC file'
   task :validate_docs do
     toc_file = ENV['TOC']
     unless toc_file
-      toc_file = File.expand_path('../../test/fixtures/toc.json', __dir__)
+      toc_file = File.expand_path('../foreman_theme_satellite/documentation-toc.json', __dir__)
```
Comment (ShimShtein) [inline lib/tasks/foreman_theme_satellite_tasks.rake:21]: This will require a change to the downstream CI

File: lib/foreman_theme_satellite/branded_words.rb [hunk 5]
```diff
@@ -53,6 +53,6 @@ module ForemanThemeSatellite
     /\b[Kk]atello 4.11\b(?!-)/       => 'Satellite 6.15',
     /\b[Kk]atello 4.12\b(?!-)/       => 'Satellite 6.16',
     /\b[Kk]atello 4.13\b(?!-)/       => 'Satellite 6.16',
-    %r{https://theforeman.org/plugins/katello/(?!-)} => "https://access.redhat.com/documentation/en-us/red_hat_satellite/#{ForemanThemeSatellite::SATELLITE_SHORT_VERSION}/html/managing_hosts/registering_hosts#registering-a-host-to-satellite-using-the-bootstrap-script"
+    %r{https://theforeman.org/plugins/katello/(?!-)} => "https://docs.redhat.com/documentation/en-us/red_hat_satellite/#{ForemanThemeSatellite::SATELLITE_SHORT_VERSION}/html/managing_hosts/registering_hosts#registering-a-host-to-satellite-using-the-bootstrap-script"
```
Comment (ShimShtein) [inline lib/foreman_theme_satellite/branded_words.rb]: Seems unrelated, can you move to a separate PR?
Also: we can use `documentation_root` here too.

File: lib/tasks/foreman_theme_satellite_tasks.rake (hunk 2 above)
Comment (ShimShtein) [reply lib/tasks/foreman_theme_satellite_tasks.rake]: It will fail if the chapter is not recognized. But what about non-recognized guide? Is it possible to have one?

File: lib/foreman_theme_satellite/documentation.rb (hunk 3 above)
Comment (ShimShtein) [reply lib/foreman_theme_satellite/documentation.rb]: The user can do nothing about this, I think we shouldn't log it as warning. It's either info or not logging it at all. 

Specifically for the tests, let's inject the logger: 
``` ruby
def self.docs_url(guide, chapter, logger:  Foreman::Logging.logger('app'))
```
So for regular work it will take the logger from the rails app (specifically Foreman logger), and for the rake task we will inject a simple ruby logger :
``` ruby
require 'logger'
docs_url('foo', 'bar', logger:  Logger.new(STDOUT))
```
I do something similar in [rh_cloud](https://github.com/theforeman/foreman_rh_cloud/blob/e939a23815de0acb3e9c321321caa7688f69f641/lib/tasks/insights.rake#L26)

**Review by ShimShtein — COMMENTED** [review]
I like the idea of using the TOC, although it's used here only for validation.

Maybe we should extend the idea to have an upstream-to-downstream mapping file, that would be  maintained automatically, so we will be sure the chapter actually exists.

The idea is to have downstream_toc.json <-> upstream_toc.json connection. The algorithm for automatic transformation would be the same as you have pointed out here plus the manual overrides. Then once one of the TOC files gets updated, we can make sure we still have valid connection.

The idea is to ensure each upstream TOC entry has an equivalent downstream: either by automatic guessing (like the algo you have suggested in this PR) or by manual override (with a dictionary). This is much easier that try to hunt down all references to documentation in the code.
WDYT?

File: lib/foreman_theme_satellite/branded_words.rb (hunk 5 above)
Comment (ekohl) [reply lib/foreman_theme_satellite/branded_words.rb]: It already is: https://github.com/RedHatSatellite/foreman_theme_satellite/pull/59.

Comment (ShimShtein) [reply lib/foreman_theme_satellite/branded_words.rb]: Commented there about the `documentation_root`

File: lib/tasks/foreman_theme_satellite_tasks.rake (hunk 2 above)
Comment (ekohl) [reply lib/tasks/foreman_theme_satellite_tasks.rake]: The `html-single` code path is the unrecognized guide. I admit it's very indirect. Note that it can only be reached if it's not found in `DOCS_GUIDE_PAGES` (which is the TOC).

File: lib/foreman_theme_satellite/documentation.rb (hunk 3 above)
Comment (ekohl) [reply lib/foreman_theme_satellite/documentation.rb]: Implemented

File: lib/tasks/foreman_theme_satellite_tasks.rake (hunk 2 above)
Comment (ShimShtein) [reply lib/tasks/foreman_theme_satellite_tasks.rake]: Let me rephrase: If I have a link without a chapter: just for the guide. It will not get into the `if chapter` block in the `docs_url` method, and will return `<root>/non-existing-guide`. Which is an error too, if I read the code correctly.

**Review by ShimShtein — COMMENTED** [review]
I'm still waiting for you opinion about the joined TOC idea.

File: lib/tasks/foreman_theme_satellite_tasks.rake (hunk 2 above)
Comment (ekohl) [reply lib/tasks/foreman_theme_satellite_tasks.rake]: There we assume `guide.downcase` returns the correct result, but you're right that we actually have a list of guides and can check. And I'll go through the list of current guides and add mappings

File: lib/foreman_theme_satellite/documentation.rb [hunk 6]
```diff
@@ -77,28 +77,38 @@ module Documentation
       'foreman_discovery' => "#{ForemanThemeSatellite.documentation_root}/provisioning_hosts/discovering-hosts-on-a-network_provisioning",
     }.freeze
 
-    DOCS_GUIDES_LINKS = {
-      'Managing_Content' => {
-        'Products_and_Repositories_content-management' => "#{ForemanThemeSatellite.documentation_root}/managing_content/importing_content_content-management#Products_and_Repositories_content-management",
-      },
+    # Guide mapping. If no entry is present, it should be assumed that it can
+    # be downcased
+    DOCS_GUIDE_MAPPING = {
```
Comment (girijaasoni) [inline lib/foreman_theme_satellite/documentation.rb:79]: It will conflict with this [change](https://github.com/RedHatSatellite/foreman_theme_satellite/pull/45/files#diff-03deddd1ea19da8194b5ae0d27b67e7804f3618c2c8ed058c78a3a8ccbf5d622R83), I suggest we rebase once the PR is merged

File: lib/tasks/foreman_theme_satellite_tasks.rake (hunk 2 above)
Comment (ekohl) [reply lib/tasks/foreman_theme_satellite_tasks.rake]: I updated the mapping based on https://github.com/theforeman/foreman-documentation/blob/master/guides/upstream_filename_to_satellite_link.json but when verifying I found some mismatches: https://github.com/theforeman/foreman-documentation/pull/3181

File: test/integration/documentation_controller_branding_test.rb [hunk 7]
```diff
@@ -7,9 +7,44 @@ def test_docs_redirect_branded
...
+    assert_redirected_to "https://docs.redhat.com/documentation/en-us/red_hat_satellite/#{ForemanThemeSatellite.documentation_version}/html-single/managing_content/index#does-not-exist"
+  end
+
   def test_docs_redirect_unknown_chapter
     get "/links/docs/Managing_Hosts"
 
     assert_redirected_to "https://docs.redhat.com/documentation/en-us/red_hat_satellite/#{ForemanThemeSatellite.documentation_version}/html/managing_hosts/"
   end
+
+  def test_many
```
Comment (ekohl) [inline test/integration/documentation_controller_branding_test.rb:45]: I'm not sure if we want to keep this, but it was a simple way for me to test if the old `DOCS_GUIDES_LINKS` were properly converted
